File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        
        map<int ,int> mp;
        vector<vector<int>> ans;

        set<vector<int>> s;
        sort(nums.begin() , nums.end());
        int n = nums.size();
        for(int i = 0 ; i < n ; i++){
            int j = i + 1;
            int k = n - 1;

            while(j < k){
                int sum = nums[i] + nums[j] + nums[k];

                if(sum == 0){
                    s.insert({nums[i] , nums[j] , nums[k]});
                    j++;
                    k--;
                }
                else if(sum > 0){
                    k--;
                }
                else{
                    j++;
                }
            }
        }

        for(auto dups : s){
            ans.push_back(dups);
        }
        return ans;
    }
};

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:

        i = 0
        n = len(nums)
        nums.sort()
        ans = []
        for i , a in enumerate(nums):
            if(a > 0):
                break
            
            if(i > 0 and nums[i - 1] == nums[i]):
                continue

            j = i + 1
            k = n - 1

            while j < k:
                if(nums[i] + nums[j] + nums[k] == 0):
                    ans.append([nums[i] , nums[j] , nums[k]])
                    j += 1
                    k -= 1
                    while(nums[j] == nums[j - 1] and j < k):
                        j += 1
                elif(nums[i] + nums[j] + nums[k] < 0):
                    j += 1
                
                else:
                    k -= 1

                
                
                
                # while(nums[k] == nums[k - 1] and j < k):
                #     k -= 1
            
           

        return ans

